documentation/correctness-previous.adoc

```diff
@@ -0,0 +1,130 @@
+= Model correctness
+
+We can identify a number of separate criteria for the correctness of a model.
+Failure to comply to one or more of these criteria doesn't imply the model as a whole is completely incorrect.
+Typically, it's quite possible to recover from such failures.
+The criteria can be visualized as follows.
+
+.The criteria for model correctness as a diagram
+image::../resources/correctness-diagrams/criteria.svg[align="center"]
+
+The discussion of these criteria below is divided into two layers:
+
+. JSON serialization of the model: the _serialization chunk_
+. The model (post-deserialization/in-memory)
+
+Each criterion has three aspects:
+
+. What are the conditions to comply to?
+. How is an incorrectness to be reported?
+. Can an incorrectness be recovered from (and if so: how)?
+
+
+== Serialization chunk
+
+=== Low-level syntactic
+
+The serialization is parsable as JSON.
+Note that “parsable” doesn't necessarily mean “syntax-correct”.
+E.g., trailing commas are not syntax-correct, but don't necessarily prevent the JSON from being parsable, nor deserializable.
+It's desirable to be able to recover from "`simple`"/"`obvious`" syntax errors in the JSON.
+
+Incorrectness is to be reported in terms of the location where parsing the JSON text failed, including an indication whether this was recoverable.
+The reporting is typically dependent on the precise JSON parser used.
+In particular, the JSON parser used might recover automatically, essentially ignoring syntax errors without reporting them in any way.
+This could be problematic, e.g. in case of duplicate keys in an object: the JSON parser is most likely to only reproduce the last value for the duplicate key, ignoring the previous ones.
+
+=== Structural syntactic
+
+The JSON conforms to https://github.com/LionWeb-io/specification/blob/main/serialization/serialization.adoc[the specification for the JSON serialization format].
+Note that JSON Schema isn't powerful enough to express more high-level, nor constraints of a referential nature on the serialization chunk.
+
+Incorrectness is to be reported in terms of e.g. a https://github.com/json-path/JsonPath[JsonPath] to the location(s) of failure(s).
```

Comment:
Does this mean that different errors can be reported in a different way? Some specifying the position and some specifying the path?
I think it would be convenient if all of the issues were reported with the position, so that they  can be highlighted in the editor by appropriate plugins. For this it would be useful to have the position as a range of characters (i.e., a starting and an ending point in the file, not a single point).
If some errors specify _also_ a JsonPath, that would be useful

Reply:
Yup. (It's also taking a bead from our own practice of designing DSLs, I think.)

Gradually “adding on” more information sounds like a good idea. Providing range information should be possible as long as we can use a JSON parser that reports that information per JSON element.

Reply:
I wrote quite some additional text on how to report a location.